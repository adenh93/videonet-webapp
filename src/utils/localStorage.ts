import { FilmDetails, SearchResult } from "../graphql/types";

type LocalStorageCollection = "WATCHED_ITEMS" | "LIKED_ITEMS";

export const setItem = <T>(type: LocalStorageCollection, item: T): void =>
  localStorage.setItem(type, JSON.stringify(item));

export const getItems = <R>(type: LocalStorageCollection): R[] => {
  const items = localStorage.getItem(type);
  if (!items) return [];
  return JSON.parse(items) as R[];
};

export const getWatchList = (): SearchResult[] =>
  getItems<SearchResult>("WATCHED_ITEMS");

export const isInWatchList = (movieId: number): boolean =>
  getWatchList().find((m) => m.id === movieId) !== undefined;

export const addToWatchList = (movie: FilmDetails): void => {
  const movies = getWatchList();
  if (!movies.find((m) => m.id === movie.id))
    setItem("WATCHED_ITEMS", [...movies, movie]);
};

export const removeFromWatchList = (movieId: number): void =>
  setItem(
    "WATCHED_ITEMS",
    getWatchList().filter((m) => m.id !== movieId)
  );

export const getLikedMovies = (): number[] => getItems<number>("LIKED_ITEMS");

export const isLiked = (movieId: number): boolean =>
  getLikedMovies().find((item) => item === movieId) !== undefined;

export const likeMovie = (movieId: number): void => {
  const likedItems = getLikedMovies();
  if (!likedItems.find((item) => item === movieId))
    setItem("LIKED_ITEMS", [...likedItems, movieId]);
};

export const unlikeMovie = (movieId: number): void => {
  setItem(
    "LIKED_ITEMS",
    getLikedMovies().filter((item) => item !== movieId)
  );
};

export const toggleWatched = (movie: FilmDetails): boolean => {
  const watched = isInWatchList(movie.id);
  watched ? removeFromWatchList(movie.id) : addToWatchList(movie);
  return !watched;
};

export const toggleLiked = (movieId: number): boolean => {
  const liked = isLiked(movieId);
  liked ? unlikeMovie(movieId) : likeMovie(movieId);
  return !liked;
};
