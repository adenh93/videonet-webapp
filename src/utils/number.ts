import numeral from "numeral";

export const formatNumber = (n: Number): String => numeral(n).format("0,0");
