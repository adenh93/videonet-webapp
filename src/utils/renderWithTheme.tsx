import React, { ReactElement } from "react";
import renderer from "react-test-renderer";
import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import { BrowserRouter as Router } from "react-router-dom";
import theme from "../theme";

export const renderWithTheme = (component: ReactElement) =>
  renderer.create(
    <ThemeProvider theme={theme}>
      <Router>{component}</Router>
    </ThemeProvider>
  );

export const rtlWithTheme = (component: ReactElement) =>
  render(
    <ThemeProvider theme={theme}>
      <Router>{component}</Router>
    </ThemeProvider>
  );
