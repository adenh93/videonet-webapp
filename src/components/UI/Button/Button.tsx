import React from "react";
import { ButtonContainer, StyledButton, ButtonIcon } from "./Styles";

export type ButtonColor = "primary" | "secondary" | "success";

interface ButtonProps extends React.HTMLAttributes<HTMLButtonElement> {
  icon?: string;
  color?: ButtonColor;
  outline?: boolean;
}

const Button: React.FC<ButtonProps> = ({
  icon,
  color = "primary",
  outline = false,
  ...props
}) => (
  <ButtonContainer>
    {icon && <ButtonIcon src={icon} />}
    <StyledButton
      icon={icon}
      color={color}
      outline={outline}
      {...props}
    ></StyledButton>
  </ButtonContainer>
);

export default Button;
