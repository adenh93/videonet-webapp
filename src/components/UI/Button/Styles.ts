import styled from "styled-components";
import { ButtonColor } from "./Button";

export const ButtonContainer = styled.div`
  position: relative;
`;

interface StyledButtonProps {
  icon?: string;
  color: ButtonColor;
  outline: boolean;
}

export const StyledButton = styled.button<StyledButtonProps>`
  color: ${({ color, outline, theme }) =>
    outline ? theme.palette.button[color] : theme.palette.text.main};
  font-size: ${({ theme }) => theme.fontSize.button};
  cursor: pointer;
  width: 18rem;
  height: 6rem;
  border-radius: ${({ theme }) => theme.borderRadius.rounded};
  background: ${({ color, outline, theme }) =>
    outline ? "none" : theme.palette.button[color]};
  border: ${({ color, outline, theme }) =>
    outline ? `2px solid ${theme.palette.button[color]}` : "none"};
  filter: drop-shadow(${({ theme }) => theme.shadows.dark});
  padding-left: ${({ icon }) => (icon ? "10%" : 0)};
  transition: transform 0.5s;

  &:disabled {
    cursor: not-allowed;
    border: ${({ color, outline, theme }) =>
      outline ? `2px solid ${theme.palette.button.disabled}` : "none"};
    background: ${({ color, outline, theme }) =>
      outline ? "none" : theme.palette.button.disabled};
    color: ${({ theme }) => theme.palette.button.disabled};
  }

  &:hover:not(:disabled) {
    transform: scale(1.05);
  }
`;

export const ButtonIcon = styled.img`
  position: absolute;
  left: 10%;
  top: 25%;
`;
