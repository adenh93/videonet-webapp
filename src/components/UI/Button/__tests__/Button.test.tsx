import React from "react";
import { renderWithTheme } from "../../../../utils/renderWithTheme";
import Button from "../Button";
import "jest-styled-components";

describe("Button UI component", () => {
  it("renders correctly", () => {
    const tree = renderWithTheme(<Button>Test</Button>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
