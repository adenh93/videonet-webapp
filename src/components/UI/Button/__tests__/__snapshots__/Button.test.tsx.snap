// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button UI component renders correctly 1`] = `
.c0 {
  position: relative;
}

.c1 {
  color: #f9f9f9;
  font-size: 2rem;
  cursor: pointer;
  width: 18rem;
  height: 6rem;
  border-radius: 20rem;
  background: #BA68C8;
  border: none;
  -webkit-filter: drop-shadow(0.6rem 0.6rem 0.6rem rgba(0,0,0,0.25));
  filter: drop-shadow(0.6rem 0.6rem 0.6rem rgba(0,0,0,0.25));
  padding-left: 0;
}

<div
  className="c0"
>
  <button
    className="c1"
    color="primary"
  >
    Test
  </button>
</div>
`;
